infra/src/main/kotlin/fi/fta/geoviite/infra/geocoding/GeocodingDao.kt

```diff
@@ -35,35 +57,100 @@ class GeocodingDao(
     fun getGeocodingContextCacheKey(
         publishType: PublishType,
         trackNumberId: IntId<TrackLayoutTrackNumber>,
+        kmPostIdsToPublish: List<IntId<TrackLayoutKmPost>>? = null,
     ): GeocodingContextCacheKey? {
-        //language=SQL
+        assert(!(publishType == PublishType.OFFICIAL && kmPostIdsToPublish != null)) {
+            "Trying to get geocoding context cache key for in OFFICIAL mode but intending to publish km post ids"
+        }
+        // if the track has been saved at all, it will have a change time, and if not, it can't have any km posts
+        // referencing it -> early exit is correct
+        val maxChangeTimeOnTrack = getTrackNumberOrReferenceLineChangeTime(publishType, trackNumberId)
+            ?: return null
+        return if (kmPostIdsToPublish != null) {
+            publicationGeocodingContextCacheKey(trackNumberId, kmPostIdsToPublish, maxChangeTimeOnTrack)
+        } else {
+            ordinaryGeocodingContextCacheKey(trackNumberId, publishType, maxChangeTimeOnTrack)
+        }
+    }
+
+    private fun ordinaryGeocodingContextCacheKey(
+        trackNumberId: IntId<TrackLayoutTrackNumber>,
+        publishType: PublishType,
+        maxChangeTimeOnTrack: Instant
+    ): OrdinaryGeocodingContextCacheKey {
+        val sql = """
+            select max(km_post.change_time) as change_time
+            from layout.km_post_version km_post
+            where km_post.track_number_id = :track_number_id
+              and ((km_post.draft and :publication_type = 'DRAFT')
+                    or (not km_post.draft and :publication_type = 'OFFICIAL'))
```

Comment:
Tässä hoidettu vielä erikseen pois se ongelma edelliseltä geokoodauskontekstien aikaleimoihin koskemiselta, että se vei draf-tilankin aikaleimoja eteenpäin olioita julkaistessa.